File: problem7.cpp
#include <iostream>
#include <cstring>
#include <cmath>
using namespace std;
const
	int limit = 1000000;
	int ans = 10001;
bool flag[limit];
void makelist()
{
	memset(flag, true, sizeof(flag));
	for (int i = 2; i < sqrt(limit); ++ i)
	{
		while (!flag[i]) ++ i;
		for (int j = 2; i * j <= limit; ++ j) flag[i * j] = false;
	}
}

int main()
{
	makelist();
	int total = 0;
	for (int i = 2; i < limit; ++ i)
		if (flag[i])
		{
			++ total;
			if (total == ans)
			{
				cout << i << endl;
				return 0;
			}
		}
	return 0;
}

/*
Answer = 104743
*/

File: problem1-50/problem7/problem7.pas
program problem7;
const maxn=1000000;
      ans=10001;
var a:array[1..maxn] of boolean;
    i,total:longint;
procedure makelist;
 var i,j:longint;
 begin
     fillchar(a,sizeof(a),true);
     i:=2; a[1]:=false;
     while i<sqrt(maxn) do begin
	 while not a[i] do inc(i);
	 j:=2;
	 while (i*j<=maxn) do begin
	     a[i*j]:=false;
	     inc(j);
	 end;
	 inc(i);
     end;
 end;

begin
    makelist;
    total:=0;
    for i:=1 to maxn do
	if a[i] then begin
	    inc(total);
	    if total=ans then begin
		writeln(i);
		halt;
	    end;
	end;
end.
